3rdparty/openvx/include/ivx.hpp

```diff
@@ -0,0 +1,1606 @@
+        if (type != vxTypeEnum) throw WrapperError("incompatible reference type");
+    }
+
+    ~RefWrapper()
+    { release(); }
+};
+
+#ifdef IVX_USE_CXX98
+
+    #define IVX_REF_STD_CTORS_AND_ASSIGNMENT(Class) \
+        Class() : RefWrapper() {} \
+        explicit Class(Class::vxType _ref, bool retainRef = false) : RefWrapper(_ref, retainRef) {} \
+        Class(const Class& _obj) : RefWrapper(_obj) {} \
+        \
+        Class& operator=(Class _obj) { using std::swap; swap(ref, _obj.ref); return *this; }
+
+#else
+
+    #define IVX_REF_STD_CTORS_AND_ASSIGNMENT(Class) \
+        Class() : RefWrapper() {} \
+        explicit Class(Class::vxType _ref, bool retainRef = false) : RefWrapper(_ref, retainRef) {} \
+        Class(const Class& _obj) : RefWrapper(_obj) {} \
+        Class(Class&& _obj) : RefWrapper(std::move(_obj)) {} \
+        \
+        Class& operator=(Class _obj) { using std::swap; swap(ref, _obj.ref); return *this; }
+
+#endif // IVX_USE_CXX98
+
+#endif // IVX_USE_EXTERNAL_REFCOUNT
+
+/*
+* Context
+*/
+class Context : public RefWrapper<vx_context>
+{
+public:
+
+    IVX_REF_STD_CTORS_AND_ASSIGNMENT(Context)
+
+    static Context create()
+    { return Context(vxCreateContext()); }
+
+    template <typename T>
+    static Context getFrom(const T& ref)
+    {
+        vx_context c = vxGetContext(castToReference(ref));
+        // vxGetContext doesn't increment ref count, let do it in wrapper c-tor
+        return Context(c, true);
+    }
+
+    void loadKernels(const std::string& module)
+    { IVX_CHECK_STATUS( vxLoadKernels(ref, module.c_str()) ); }
+};
+
+/*
+* Graph
+*/
+class Graph : public RefWrapper<vx_graph>
+{
+public:
+    IVX_REF_STD_CTORS_AND_ASSIGNMENT(Graph);
+
+    static Graph create(vx_context c)
+    { return Graph(vxCreateGraph(c)); }
+
+    void verify()
+    { IVX_CHECK_STATUS( vxVerifyGraph(ref) ); }
+
+    void process()
+    { IVX_CHECK_STATUS( vxProcessGraph(ref) ); }
+
+    void schedule()
+    { IVX_CHECK_STATUS(vxScheduleGraph(ref) ); }
+
+    void wait()
+    { IVX_CHECK_STATUS(vxWaitGraph(ref)); }
+};
+
+/*
+* Kernel
+*/
+class Kernel : public RefWrapper<vx_kernel>
+{
+public:
+    IVX_REF_STD_CTORS_AND_ASSIGNMENT(Kernel);
+
+    static Kernel getByEnum(vx_context c, vx_enum kernelID)
+    { return Kernel(vxGetKernelByEnum(c, kernelID)); }
+
+    static Kernel getByName(vx_context c, const std::string& name)
+    { return Kernel(vxGetKernelByName(c, name.c_str())); }
+};
+
+/*
+* Node
+*/
+#ifdef IVX_USE_CXX98
+
+class Node : public RefWrapper<vx_node>
+{
+public:
+    IVX_REF_STD_CTORS_AND_ASSIGNMENT(Node);
+
+    static Node create(vx_graph g, vx_kernel k)
+    { return Node(vxCreateGenericNode(g, k)); }
+
+    static Node create(vx_graph graph, vx_kernel kernel, const std::vector<vx_reference>& params)
+    {
+        Node node = Node::create(graph, kernel);
+        vx_uint32 i = 0;
+        for (std::vector<vx_reference>::const_iterator p = params.begin(); p != params.end(); p++)
```

Comment:
I don't remember this note in C++98 and other standards (especially for complex types with overloaded operators), so lets follow BKM for general case:

> Because a temporary copy of the object is constructed during post-increment and post-decrement, pre-increment or pre-decrement operators are usually more efficient in contexts where the returned value is not used.

http://en.cppreference.com/w/cpp/language/operator_incdec